splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+                'invalid server state: INQUIRE not originating from gpg-agent')
+        loop = asyncio.get_event_loop()
+        self.inquire_complete = loop.create_future()
+        self.inquire_commands = inquire_commands
+        self.state = ServerState.client_inquire
+        self.client_write(b'INQUIRE ' + inquire + b'\n')
+        await self.inquire_complete
+        self.inquire_commands = {}
+        self.state = ServerState.agent_response
+
+    def fake_respond(self, response):
+        self.client_write(response + b'\n')
+
+    @staticmethod
+    def verify_keygrip_arguments(min_count, max_count, untrusted_args):
+        args_regex = re.compile(
+            rb'[0-9A-F]{40}( [0-9A-F]{40}){%d,%d}' % (min_count-1, max_count-1))
+
+        if args_regex.fullmatch(untrusted_args) is None:
+            raise Filtered
+        return untrusted_args
+
+    def sanitize_key_desc(self, untrusted_args: bytes) -> bytes:
+        untrusted_args = untrusted_args.replace(b'+', b' ')
+        untrusted_args = re.sub(
+            rb'%[0-9A-F]{2}',
+            lambda m: bytes([int(m.group(0)[1:].decode('ascii'), 16)]),
+            untrusted_args,
+            flags=re.IGNORECASE
+        )
+        allowed_ascii = (string.printable + '\x0a').encode('ascii')
+        args = "Message from '{}':\n{}".format(
+            self.client_domain,
+            ''.join((chr(c) if c in allowed_ascii else '.')
+                    for c in untrusted_args)
+        )
+        return args.replace('%', '%25').\
+            replace('+', '%2B').\
+            replace('\n', '%0A').\
+            replace(' ', '+').\
+            encode('ascii')
+
+    async def command_RESET(self, untrusted_args: Optional[bytes]):
+        if untrusted_args:
+            raise Filtered
+        await self.agent_protocol.send_command(b'RESET', None)
+
+    async def command_OPTION(self, untrusted_args: Optional[bytes]):
+        if not untrusted_args:
+            raise Filtered
+
+        try:
+            untrusted_name, untrusted_value = untrusted_args.split(b'=', 1)
+        except ValueError:
+            untrusted_name = untrusted_args
+            untrusted_value = None
+        try:
+            action, opts = self.options[untrusted_name]
+            name = untrusted_name
+        except KeyError:
+            raise Filtered
+
+        if action == OptionType.override:
+            if opts:
+                option_arg = b'%s=%s' % (name, opts)
+            else:
+                option_arg = name
+        elif action == OptionType.verify:
+            if callable(opts):
+                verified = opts(untrusted_value=untrusted_value)
+            elif isinstance(opts, Pattern):
+                verified = (opts.fullmatch(untrusted_value) is not None)
+            else:
+                verified = (untrusted_value == opts)
+            if not verified:
+                raise Filtered
+            value = untrusted_value
+
+            if value:
```

Comment:
`is not None`

Reply:
Should really empty values be allowed? (thing like `OPTION something=`)

Reply:
I think we should not "silently" change the form. If there will be ever a case where we want to allow "" this is another question. Currently we never allow "".